grpc-client-server/README.md

```diff
@@ -0,0 +1,188 @@
+message ProbeRequest {
+  string name = 1;
+}
+
+// The response message containing the greetings
+message ProbeReply {
+  string message = 1;
+}
+```
+
+Regenerate the generated protobuf code: 
+```
+protoc --go_out=. --go_opt=paths=source_relative \
+    --go-grpc_out=. --go-grpc_opt=paths=source_relative prober/prober.proto
+```
+
+Observe the new generated files:
+```
+ls prober
+prober.pb.go		prober.proto		prober_grpc.pb.go
+```
+
+Read through `prober_grpc.pb.go` - this is the interface you will use in your code. This is how gRPC works: a `proto` format 
+gets generated into language-specific code that you can use to interact with gRPCs. If you're working with multiple programming languages
+that need to interact through RPCs, you can do this by generating from the same protocol buffer definition with the language-specific
+tooling. 
+
+Now, update the client and server `main.go` files to use the changed package, service and message names.
+Get these running again as they did originally. The functionality will still be the same for now, just the hello world interaction.
+
+```
+go run prober_server/main.go
+2022/10/09 20:20:55 server listening at [::]:50051
+2022/10/09 20:21:24 Received: world
+````
+
+```
+go run prober_client/main.go 
+2022/10/09 20:21:24 Greeting: Hello world
+```
+
+While we haven't yet changed what the code does, we've gained some experience with the protocol buffer format, learned 
+how to generate Go code from protocol buffer definitions, and call that code from Go programs.
+
+Note: in a real production system you can't rename protocol buffers as we've done here: you would run into problems because your old clients can't talk to your new servers, and vice-versa. Since we gradually release code, we would have to support both services until all old clients were turned down.
+
+## Implement prober logic
+
+Let's modify the prober service slightly. Instead of the simple HelloWorld greeting, we are going to implement a service that will probe a HTTP endpoint N times and return the average time to fetch that endpoint to the client. 
+
+This tool could be useful in a real operational context. Imagine that we want to verify that our site is available and has acceptable latency from many different locations around the world. We might run several instances of our prober server in
+different regions. 
+
+Change your prober request and response to look like this and regenerate your generated go code.
+```
+// The request message
+message ProbeRequest {
+  string endpoint = 1;    // Endpoint to probe
+  int32 repetitions = 2;  // How many probes to execute
+}
+
+// The response message 
+message ProbeReply {
+  float average = 1;    // Average response time (in milliseconds)
+}
+```
+
+Update your client to read the endpoint and number of repetitions from the [command line](https://gobyexample.com/command-line-arguments).
+Then update your server to execute the probe: do a HTTP fetch of `endpoint` the specified number of times.
```

Comment:
fetch → `GET` IMO

"do a HTTP fetch of `endpoint` the specified number of times" is also quite colloquial — possibly not clear to a non-native speaker.